### ISSUE TITLE
Call chain death (unresolved promise) during connect() init-stage errors

### ISSUE THREAD

**Issue by janakaud** [theophilusx/ssh2-sftp-client#426]
(I went through recent existing issues but could not find one matching my scenario; apologies and feel free to link/close this, if it has already been reported.)

On 9.0.2, it seems that if an error occurs before connection creation begins (e.g. during private key parsing), the call chain just "dies"; neither does respective error get propagated to the caller, nor does the post-connection callback (promise `then()`/`catch()` from `sftp.connect()` call) get invoked.

Earlier I was on 6.0.1, which did not have this problem (with non-RSA keys I got the expected "Cannot parse privateKey: Unsupported key format" error, in `sftp.connect()`'s `catch()`).

From what I understand so far, the problem is with `SFTP.end()`, where the returned promise is only resolved during an actual connection closure (in `endCloseHandler` callback); if the internal `client` does not yet have a connection (socket), this callback seems not getting called, so the flow ends up with an unresolved promise:

`src/index.js`, class `SftpClient`
```
  end() {
    ..
    return new Promise((resolve, reject) => {
      listeners = addTempListeners(this, 'end', reject);
      this.endCalled = true;
      endCloseHandler = () => {
        this.sftp = undefined;
        this.debugMsg('end: Connection closed');
        resolve(true);
      };
      this.on('close', endCloseHandler);
      if (this.client) {
        this.client.end();	// <-- during initial stages, won't invoke endCloseHandler
      } else {
        // no actual connection exists - just resolve
        this.debugMsg('end: Called when no connection active');
        resolve(true);
      }
    ..
  }
```

I am not entirely sure how to handle this properly; currently my idea is that the promise should be resolved immediately if `this.client` is "not connected", instead of setting up `endCloseHandler` callback - but I could not find an API method on the underlying `ssh2` client, to check its current status (i.e. whether it is currently "connected" or "connecting").

FWIW currently I'm using the following hack on the caller's side, to bypass the internal `sftp.end()` logic entirely if the internal `sftp.client` does not have a `_sock`. This at least seems to cover some basic cases like "Unsupported key format" above, although it might fail where the client does have a socket object, but it is not actually "alive" (suspicion based on currently used (1.11.0) implementation of `ssh2`'s `Client.end()`).

```
    const sftp = new SFTP();
    const origEndFn = sftp.end.bind(sftp);
    sftp.end = function() {
        if (sftp.client && !sftp.client._sock) {
            return;
        }
        return origEndFn();
    };
    sftp.connect(connConfig).then(..
```

Any ideas on a better resolution?

**Comment by janakaud** [comment]
SFTP debug logs with an "unsupported" (non-RSA key), 9.0.2:

```
CLIENT[sftp]: connect: Debugging turned on
CLIENT[sftp]: ssh2-sftp-client Version: 9.0.2  {
 "node": "16.13.2",
 "v8": "9.4.146.24-node.14",
 "uv": "1.42.0",
 "zlib": "1.2.11",
 "brotli": "1.0.9",
 "ares": "1.18.1",
 "modules": "93",
 "nghttp2": "1.45.1",
 "napi": "8",
 "llhttp": "6.0.4",
 "openssl": "1.1.1l+quic",
 "cldr": "39.0",
 "icu": "69.1",
 "tz": "2021a",
 "unicode": "13.0",
 "ngtcp2": "0.1.0-DEV",
 "nghttp3": "0.1.0-DEV"
}
CLIENT[sftp]: connect: Connect attempt 1

<no further output; my sample script only had a connect().then(..).catch(..), and it just exited at this point>
```

Same config, with 6.0.1:

```
CLIENT[sftp]: Debugging turned on
CLIENT[sftp]: Connect attempt 1
CLIENT[sftp]: Removing listener from ready
CLIENT[sftp]: Connect attempt 2
CLIENT[sftp]: Removing listener from ready
catch Error: Cannot parse privateKey: Unsupported key format
    at Client.connect (C:\Users\janaka\code\dustbin\node-sftp\client-6.0.1\node_modules\ssh2\lib\client.js:243:13)
    at C:\Users\janaka\code\dustbin\node-sftp\client-6.0.1\node_modules\ssh2-sftp-client\src\index.js:117:45
    at new Promise (<anonymous>)
    at SftpClient.sftpConnect (C:\Users\janaka\code\dustbin\node-sftp\client-6.0.1\node_modules\ssh2-sftp-client\src\index.js:103:12)
    at promiseRetry.retries (C:\Users\janaka\code\dustbin\node-sftp\client-6.0.1\node_modules\ssh2-sftp-client\src\index.js:143:21)
    at C:\Users\janaka\code\dustbin\node-sftp\client-6.0.1\node_modules\promise-retry\index.js:29:24 undefined
```

FWIW, sample script used for above:

```
const SFTP = require('ssh2-sftp-client');
const fs = require('fs');

try {
  new SFTP().connect({
    debug: msg => console.log(msg),
    retries: 0,
    host: 'redacted',
    port: 22,
    username: 'redacted',
    privateKey: fs.readFileSync('redacted')
  })
    .then((r1, r2) => console.log('then', r1, r2))
    .catch((e1, e2) => console.error('catch', e1, e2));
} catch (e) {
  console.error('try-catch', e);
}
```

**Comment by theophilusx** [comment]
> From what I understand so far, the problem is with SFTP.end(), where the returned promise is only resolved during an
> actual connection closure (in endCloseHandler callback); if the internal client does not yet have a connection (socket), this
> callback seems not getting called, so the flow ends up with an unresolved promise:
>
> src/index.js, class SftpClient
>
>   end() {
>     ..
>     return new Promise((resolve, reject) => {
>       listeners = addTempListeners(this, 'end', reject);
>       this.endCalled = true;
>       endCloseHandler = () => {
>         this.sftp = undefined;
>         this.debugMsg('end: Connection closed');
>         resolve(true);
>       };
>       this.on('close', endCloseHandler);
>       if (this.client) {
>         this.client.end();	// <-- during initial stages, won't invoke endCloseHandler
>       } else {
>         // no actual connection exists - just resolve
>         this.debugMsg('end: Called when no connection active');
>         resolve(true);
>       }
>     ..
>   }
>

The ssh2 sftp.end() method only gets called if the this.client property
is defined. That property is only defined when a connection has been
established. If no sftp connection has been established, then the else
branch is followed, which simply resolves the promise.

Now it is possible that ssh2 is doing something 'unusual' and not
calling a close event, but that will need more investigation to verify. 

Note that there is little point in making any comparison with the older
6.x version as the underlying ssh2 module has been completely re-written
since v6.x was released (v8 onwards are all based on ssh2 v1.x, which is
a complete re-write). Error messages and emitted eventgs are very
different between the two versions. 

> I am not entirely sure how to handle this properly; currently my idea is that the promise should be resolved immediately if
> this.client is "not connected", instead of setting up endCloseHandler callback - but I could not find an API method on the
> underlying ssh2 client, to check its current status (i.e. whether it is currently "connected" or "connecting").
>

The close handler should never come into play if the connection was not
established because the 'client' property will be undefined. Note also
that there are default end() and close() handlers which are always
'active' that will set the client property to undefined if a end or
close event is fired an no other handler deals wiht it (so even if ssh2
created a connection and then disconnectgs it before end() is called,
this should still work). 

> FWIW currently I'm using the following hack on the caller's side, to bypass the internal sftp.end() logic entirely if the
> internal sftp.client does not have a _sock. This at least seems to cover some basic cases like "Unsupported key format"
> above, although it might fail where the client does have a socket object, but it is not actually "alive" (suspicion based on
> currently used (1.11.0) implementation of ssh2's Client.end()).
>
>     const sftp = new SFTP();
>     const origEndFn = sftp.end.bind(sftp);
>     sftp.end = function() {
>         if (sftp.client && !sftp.client._sock) {
>             return;
>         }
>         return origEndFn();
>     };
>     sftp.connect(connConfig).then(..
>
> Any ideas on a better resolution?
>

That 'work around' should not be necessary as the actual call to
this.client.end() should only occur when this.client is defined and
this.client is only defined if a connection has been created and ssh2
has not emitted either an end or close event. If this.client is defined
and there is no connection, then either ssh2 has failed to emit an
end/cloe event when it should or something else outside ssh2 has killed
the connection and not been registered by ssh2 and so no end or close
event emitted (in which case ssh2 would still have a _sock property as well).

So basically, I think there is something else going on here. First step
will be to have a script and reliable way to reproduce the issue along
with detailed debugging logs. It is possible ssh2 is not emitting a
close/end event when it should (I already have another issue logged with
ssh2 regarding missing close/end events for read/write streams).

**Comment by janakaud** [comment]
Thanks Tim for the detailed explanation!

Agreed on all points, except this one:

> [this.client] That property is only defined when a connection has been established.
> if the connection was not established because the 'client' property will be undefined

From what I see, in 9.0.2, `this.client` is initialized in the `SftpClient` *constructor* itself:

```
class SftpClient {
  constructor(clientName) {
    this.version = '9.0.2';
    this.client = new Client();
    this.sftp = undefined;
    ..
```

> .. default end() and close() handlers which are always 'active' that will set the client property to undefined if a end or close event is fired an no other handler deals with it (so even *if ssh2 created a connection and then disconnects it* before end() is called, this should still work).

As I understand, in my case the process fails *very early* (while `ssh2`'s `Client.connect(cfg)` is *still gathering configurations* in order to create the connection; to be precise, in the private-key-parsing `parseKey` stage). FWIW, this is the original stacktrace of the error in my case:

```
Error: Cannot parse privateKey: Unsupported key format
    at Client.connect (..\node_modules\ssh2\lib\client.js:261:15)
    at ..\node_modules\ssh2-sftp-client\src\index.js:160:19
    at new Promise (<anonymous>)
    at SftpClient.getConnection (..\node_modules\ssh2-sftp-client\src\index.js:153:12)
    at ..\node_modules\ssh2-sftp-client\src\index.js:218:22
    at ..\node_modules\promise-retry\index.js:29:24
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async SftpClient.connect (..\node_modules\ssh2-sftp-client\src\index.js:215:7)
```

I have also verified the existence of `this.client` (in a rather crude way, by patching a `console.log(this.client)` into the `if` branch of `SftpClient.end()`):

```
  end() {
      ..
      this.on('close', endCloseHandler);
      if (this.client) {
console.log('end client', this.client);   // <-- my repro. script reaches this, and logs: "end client <ref *1> Client {  _events: [Object: null prototype] { ..."
        this.client.end();
      } else {
        ..
```

> First step will be to have a script and reliable way to reproduce the issue along with detailed debugging logs.

Agreed! As of now, I can reliably (always) reproduce this using the sample script included at the end of my previous comment (second comment on this thread, which starts with sample outputs); let me know if I can provide the script in a more useful manner (zipfile, GitHub repo, etc.). Right now I run it under a vanilla environment, with only the `"ssh2-sftp-client": "9.0.2"` dependency installed.

**Comment by theophilusx** [comment]
I have just realised I have had a major brain fart!

Your sort of right and I made a mistake when I changed the end()
function (it was changed between 9.0.1 and 9.0.2.

The issue is the test is wrong. It should be

if (this.client.sftp) {
  ...
} else {
  ...
}

and not

if (this.client) {
  ...
} else {
  ...
}

This shows I need to add some additional unit tests to verify the else
branch more effectively.

It was your demonstration of when the client property is set which made
me realise my mistake, so thanks.

I will try to find time to make the change and push an update in the
next day or so.
